File: leetcode/hash sets/1496. Path Crossing/golang/main.go
package main

import (
	"fmt"
)

func isPathCrossing(path string) bool {
	type point struct {
		x int
		y int
	}

	visCoords := make(map[point]struct{})

	x, y := 0, 0

	visCoords[point{x, y}] = struct{}{}

	for _, step := range path {
		switch step {
		case 'N':
			y++
		case 'S':
			y--
		case 'E':
			x++
		case 'W':
			x--
		}

		curPos := point{x, y}

		if _, exists := visCoords[curPos]; exists {
			return true
		}

		visCoords[curPos] = struct{}{}

	}
	return false
}

func main() {
	path := "NESWW"
	fmt.Println(isPathCrossing(path))
}

File: leetcode/hash sets/1496. Path Crossing/python/main.py
class Solution:
    def isPathCrossing(self, path: str) -> bool:
        x, y = 0, 0
        vis_coords = {(x, y)}

        for step in path:
            print(step)
            match step:
                case "N":
                    dx, dy = 0, 1
                case "S":
                    dx, dy = 0, -1
                case "E":
                    dx, dy = 1, 0
                case "W":
                    dx, dy = -1, 0
                case _:
                    return False

            x += dx
            y += dy
            cur_coords = (x, y)

            if cur_coords in vis_coords:
                return True
            vis_coords.add(cur_coords)

        return False


path = "NESWW"
print(Solution().isPathCrossing(path))
